orttraining/orttraining/eager/opgen/opgen/generator.py

```diff
@@ -472,27 +522,40 @@ def _write_function_body(self, writer: writer.SourceWriter, mapped_func: MappedO
 
         if len(in_place_params) == 0:
             # tensor options
-            writer.write(f"at::TensorOptions tensor_options = {first_param.identifier.value}")
-            if first_param.parameter_type.desugar().identifier_tokens[0].value == "TensorList":
-                writer.write("[0]")
-            writer.write(".options()")
-            if need_type_promotion:
-                writer.write(".dtype(*promoted_type)")
-            writer.writeline(";")
-
-            writer.writeline("return aten_tensor_from_ort(")
-            writer.push_indent()
-            if (
-                isinstance(cpp_func.return_type, ast.TemplateType)
-                and cpp_func.return_type.identifier_tokens[-1].value == "std::vector"
-            ):
-                writer.writeline(f"{return_outputs},")
-                writer.writeline("tensor_options);")
+            if set_out_tensor:
+                if cast_output_to_bool:
+                    writer.writeline("if (out.scalar_type() != at::kBool) {")
+                    writer.push_indent()
+                    writer.writeline(f"auto temp = CastToType(invoker, {return_outputs}[0], out.scalar_type());")
+                    writer.writeline("copy(invoker, temp, ort_input_out);")
+                    writer.pop_indent()
+                    writer.writeline("}")
+
+                writer.writeline(f"return {last_param.identifier.value};")
+                return
+
             else:
-                writer.writeline(f"std::move({return_outputs}[0]),")
-                writer.writeline("tensor_options);")
-            writer.pop_indent()
-            return
+                writer.write(f"at::TensorOptions tensor_options = {first_param.identifier.value}")
+                if first_param.parameter_type.desugar().identifier_tokens[0].value == "TensorList":
```

Comment:
please leave some todo. we need better handle for TensorList in the future.

Reply:
what would the TODO suggest? The only generated op with TensorList that I found was `_cat` and it looks correct.

Reply:
here we hardcode only handle first tensor in TensorList right? we will need to extend it for more general case at some point.

Reply:
make sense. I'm going to complete this PR as all the tests are passing but will add todos for all these hardcoded param limitations in my follow-up PR.